packages/core/src/routes/experience/classes/experience-interaction.ts

```diff
@@ -0,0 +1,164 @@
+      this.verificationRecords.set(instance.type, instance);
+    }
+  }
+
+  /** Set the interaction event for the current interaction */
+  public setInteractionEvent(event: InteractionEvent) {
+    // TODO: conflict event check (e.g. reset password session can't be used for sign in)
+    this.interactionEvent = event;
+  }
+
+  /** Set the verified `accountId` of the current interaction from the verification record */
+  public identifyUser(verificationRecord: VerificationRecord) {
+    // Throws an 404 error if the user is not found by the given verification record
+    assertThat(
+      verificationRecord.verifiedUserId,
+      new RequestError({
+        code: 'user.user_not_exist',
+        status: 404,
+      })
+    );
+
+    // Throws an 409 error if the current session has already identified a different user
+    if (this.accountId) {
+      assertThat(
+        this.accountId === verificationRecord.verifiedUserId,
+        new RequestError({ code: 'session.identity_conflict', status: 409 })
+      );
+      return;
+    }
+
+    this.accountId = verificationRecord.verifiedUserId;
+  }
+
+  /**
+   * Append a new verification record to the current interaction.
+   * If a record with the same type already exists, it will be replaced.
+   */
+  public appendVerificationRecord(record: VerificationRecord) {
+    const { type } = record;
+
+    this.verificationRecords.set(type, record);
+  }
+
+  public getVerificationRecordById(verificationId: string) {
+    return this.verificationRecordsArray.find((record) => record.id === verificationId);
+  }
+
+  /** Save the current interaction result. */
+  public async save() {
+    // `mergeWithLastSubmission` will only merge current request's interaction results.
+    // Manually merge with previous interaction results here.
+    // @see {@link https://github.com/panva/node-oidc-provider/blob/c243bf6b6663c41ff3e75c09b95fb978eba87381/lib/actions/authorization/interactions.js#L106}
+
+    const { provider } = this.tenant;
+    const details = await provider.interactionDetails(this.ctx.req, this.ctx.res);
+
+    await provider.interactionResult(
+      this.ctx.req,
+      this.ctx.res,
+      { ...details.result, ...this.toJson() },
+      { mergeWithLastSubmission: true }
+    );
+  }
+
+  /** Submit the current interaction result to the OIDC provider and clear the interaction data */
+  public async submit() {
+    // TODO: refine the error code
+    assertThat(this.accountId, 'session.verification_session_not_found');
+
+    const { provider } = this.tenant;
+
+    const redirectTo = await provider.interactionResult(this.ctx.req, this.ctx.res, {
+      login: { accountId: this.accountId },
+    });
+
+    this.ctx.body = { redirectTo };
+  }
+
+  private get verificationRecordsArray() {
+    return [...this.verificationRecords.values()];
+  }
+
+  /** Convert the current interaction to JSON, so that it can be stored as the OIDC provider interaction result */
+  private toJson() {
```

Comment:
nit: maybe no need to private this. we can actually have an abstract class like "Encodable" in the future